### PR TITLE
8267836: Separate eager reclaim remembered set threshold from G1RSetSparseRegionEntries

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  currently the remembered set eager reclaim threshold is based directly on G1RSetSparseRegionEntries. I.e. otherwise eligible humongous objects are only considered candidates for eager reclaim if their remembered set has less than G1RSetSparseRegionEntries entries.

The reason for this limitation has been that before JDK-8233919 (in JDK14) it has been very expensive to get a remembered set size estimate if the remembered set for that heap region had expanded to fine tables.

This is not the case any more, and there are situation where you want to set this threshold (much) larger than the number of sparse table entries.

By default, this new value will still be based on the current value. 

Unfortunately I did not manage to create a useful test case that verifies whether a given threshold is met.

Testing: tier1

Thanks,
  Thomas

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267836](https://bugs.openjdk.java.net/browse/JDK-8267836): Separate eager reclaim remembered set threshold from G1RSetSparseRegionEntries

### Reviewers
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**)
 * [Leo Korinth](https://openjdk.java.net/census#lkorinth) (@lkorinth - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4248/head:pull/4248` \
`$ git checkout pull/4248`

Update a local copy of the PR: \
`$ git checkout pull/4248` \
`$ git pull https://git.openjdk.java.net/jdk pull/4248/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4248`

View PR using the GUI difftool: \
`$ git pr show -t 4248`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4248.diff">https://git.openjdk.java.net/jdk/pull/4248.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1Arguments.cpp [hunk 1]
```diff
@@ -49,7 +49,15 @@ void G1Arguments::initialize_alignments() {
   // size, but the heap size should be aligned with the region size. To get
   // around this we use the unaligned values for the heap.
   HeapRegion::setup_heap_region_size(MaxHeapSize);
+
+  // The remembered set needs the heap regions set up.
   HeapRegionRemSet::setup_remset_size();
+  // Needs remembered set initialization as the ergonomics are based
+  // on it.
+  if (FLAG_IS_DEFAULT(G1EagerReclaimRemSetThreshold)) {
+      guarantee(G1RSetSparseRegionEntries > 0, "");
```
Comment (lkorinth) [inline src/hotspot/share/gc/g1/g1Arguments.cpp:58]: please reduce "guarantee()" indentation and  consider not using an empty string

Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1Arguments.cpp:58]: This has been a leftover debug code. I'll remove it, sorry.

**Review by kstefanj — APPROVED** [review]
Looks good.

**Review by lkorinth — APPROVED** [review]
Looks good to me.

**Comment by tschatzl** [comment]
Thanks @lkorinth @kstefanj for your reviews.

/integrate

**Pull request closed without merging**